### PR TITLE
feat: Project Post Endpoint

### PR DESCRIPTION
**Pull request by alexanderallen-aa**
## Pre-Requisites
- [x] Yes, I updated [Authors.md](../blob/main/Authors.md) **OR** this is not my first contribution
- [x] Yes, I included and/or modified tests to cover relevent code **OR** my change is non-technical
- [x] Yes, I wrote this code entirely myself **OR** I properly attributed these changes in [Third Party Notices](../blob/main/THIRD-PARTY-NOTICES.txt)

### CODE REVIEW

File: packages/api/src/api/project/post.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  
+  const query = {id:req.user.id};
+  
+  // check for existing project
+  try {
+    const user = await em.findOne(User, query);
+    if (user?.project!==undefined) {
+      res.status(409);
+      return;
+    }
```
Comment (SpencerKaiser) [inline packages/api/src/api/project/post.ts:21]: Honestly we don't even need to make the query... if the `user` has a `project` at ALL, it means the entity exists and you should throw a 409...
```suggestion
    if (user?.project) {
      res.status(409);
      return;
    }
```

File: packages/api/src/api/project/post.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  }
+  catch {
+    res.status(500);
+    return;
+  }
+  try {
+    const project = new Project(data);
+    em.persistAndFlush(project);
```
Comment (SpencerKaiser) [inline packages/api/src/api/project/post.ts:29]:
```suggestion
    await em.persistAndFlush(project);
```

File: packages/api/src/api/project/post.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  
+  // check for existing project
+  try {
+    const user = await em.findOne(User, query);
+    if (user?.project!==undefined) {
+      res.status(409);
+      return;
+    }
+  }
+  catch {
```
Comment (SpencerKaiser) [inline packages/api/src/api/project/post.ts:23]: You can get rid of the `try ... catch` after my suggestion above

File: packages/api/src/api/project/post.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    return;
+  }
+  try {
+    const project = new Project(data);
+    em.persistAndFlush(project);
+    res.send(project);
+  }
+  catch(err) {
+    logger.error(err);
+    res.send(500);
```
Comment (SpencerKaiser) [inline packages/api/src/api/project/post.ts:34]: After creating the transaction above, you should check the error's status code and figure out if the error is an error due to the user being locked; if it is, throw a 409. [Example of how to check the status code](https://github.com/AmericanAirlines/Hangar/commit/07bfd5126b778012ed01361bbc7efa3f7e05376c#diff-c1aa0872df8ca3062b76e2375d28fa0961683b7790793fe0407ab6b113c5de8bL19-L24).

File: packages/api/src/api/project/post.ts (hunk 2 above)
Comment (SpencerKaiser) [inline packages/api/src/api/project/post.ts:29]: These two lines are the ones that need to be wrapped in a transaction; you basically need to do the following:
- start a transaction
- use **_that_** entity manager to re-fetch the user but in doing so, use a `LockMode.PESSIMISTIC_WRITE_OR_FAIL` locking mode (check me on that but I'm pretty sure that's what we want)
  - If that call fails, the transaction will fail and you'll end up in your catch block; see my comment below about how to handle that
- If the user was obtainable, create the project, persist the project, then update the user with the project, persist the user, flush both

**Review by SpencerKaiser — CHANGES_REQUESTED** [review]
You have a perfect start to it here! You're on the right path, just need to tweak how you check for a project and then add transaction support!

File: packages/api/src/api/project/post.ts (hunk 2 above)
Comment (michael20148205) [reply packages/api/src/api/project/post.ts]: []()****@ ****

Comment (michael20148205) [reply packages/api/src/api/project/post.ts]: @

**Review by granthill7 — APPROVED** [review]

**Pull request merged**